inlong-sort/sort-connectors/tubemq/src/main/java/org/apache/inlong/sort/tubemq/table/TubeMQDynamicTableFactory.java

```diff
@@ -0,0 +1,180 @@
+import static org.apache.inlong.sort.tubemq.table.TubeMQOptions.getTubeMQProperties;
+
+/**
+ * A dynamic table factory implementation for TubeMQ.
+ */
+public class TubeMQDynamicTableFactory implements DynamicTableSourceFactory {
+
+    public static final String IDENTIFIER = "tubemq";
+
+    public static final List<String> INNERFORMATTYPE = Arrays.asList("inlong-msg");
+
+    public static boolean innerFormat = false;
+
+    private static DecodingFormat<DeserializationSchema<RowData>> getValueDecodingFormat(
+            TableFactoryHelper helper) {
+        return helper.discoverOptionalDecodingFormat(DeserializationFormatFactory.class, FORMAT)
+                .orElseGet(() -> helper.discoverDecodingFormat(DeserializationFormatFactory.class, FORMAT));
+    }
+
+    private static void validatePKConstraints(
+            ObjectIdentifier tableName, CatalogTable catalogTable, Format format) {
+        if (catalogTable.getSchema().getPrimaryKey().isPresent()
+                && format.getChangelogMode().containsOnly(RowKind.INSERT)) {
+            Configuration options = Configuration.fromMap(catalogTable.getOptions());
+            String formatName = options.getOptional(FORMAT).orElse(options.get(FORMAT));
+            innerFormat = INNERFORMATTYPE.contains(formatName);
+            throw new ValidationException(String.format(
+                    "The TubeMQ table '%s' with '%s' format doesn't support defining PRIMARY KEY constraint"
+                            + " on the table, because it can't guarantee the semantic of primary key.",
+                    tableName.asSummaryString(), formatName));
+        }
+    }
+
+    private static Optional<DecodingFormat<DeserializationSchema<RowData>>> getKeyDecodingFormat(
+            TableFactoryHelper helper) {
+        final Optional<DecodingFormat<DeserializationSchema<RowData>>> keyDecodingFormat = helper
+                .discoverOptionalDecodingFormat(DeserializationFormatFactory.class, KEY_FORMAT);
+        keyDecodingFormat.ifPresent(format -> {
+            if (!format.getChangelogMode().containsOnly(RowKind.INSERT)) {
+                throw new ValidationException(String.format(
+                        "A key format should only deal with INSERT-only records. "
+                                + "But %s has a changelog mode of %s.",
+                        helper.getOptions().get(KEY_FORMAT),
+                        format.getChangelogMode()));
+            }
+        });
+        return keyDecodingFormat;
+    }
+
+    @Override
+    public DynamicTableSource createDynamicTableSource(Context context) {
+        final TableFactoryHelper helper = FactoryUtil.createTableFactoryHelper(this, context);
+
+        final ReadableConfig tableOptions = helper.getOptions();
+
+        final DecodingFormat<DeserializationSchema<RowData>> valueDecodingFormat = getValueDecodingFormat(helper);
+
+        // validate all options
+        helper.validate();
```

Comment:
Here table options with 'properties.*' will be validate and here will throw exception?
Any other source with their own properties validate their own options will use `validateExcept` to except some prefix.

Reply:
This is an extension and currently has no impact.